### DISCUSSION TITLE
Blazor server cookie authentication

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#1819] [Questions]
I have followed the approach in the Blazor book and use cookie authentication. 

I stumbled across an issue when accessing a page where user not required to log in, I receive a 401 Not authorized error in the browser console output. Page is displaying correctly, but events are not responsive causing the CSLA property validation not to work.
Looks like the signalR connection is not working and I believe this has to do with the authentication.
When I do login and then access the page, then everything is working fine.

I know this is not strictly CSLA related, but hopefully someone knows how handle this use case where I like to access certain blazor server pages when user not authenticated?

**Comment by JacoJordaan** [comment]
It seems like this is csla authorization related.

Commenting out `services.AddCsla().WithBlazorServerSupport();` in Startup.cs will NOT throw the following error:
````
blazor.server.js:1 [2020-09-23T12:36:58.589Z] Error: Failed to complete negotiation with the server: Error: Unauthorized
e.log @ blazor.server.js:1
````

**Reply by rockfordlhotka** [reply]
Hmm. 

The `UseCsla` method does this:

```c#
    public static IApplicationBuilder UseCsla(
      this IApplicationBuilder app, Action<CslaConfiguration> config)
    {
      ApplicationContext.DefaultServiceProvider = app.ApplicationServices;
      ApplicationContext.WebContextManager = 
        new Csla.AspNetCore.ApplicationContextManager(app.ApplicationServices);
      config?.Invoke(CslaConfiguration.Configure());
      return app;
    }
```

I don't think anything in here would impact authentication. It sets up the DI container and context manager, and invokes any CSLA configuration you've provided.

The `WithBlazorServerSupport` method does this:

```c#
    public static ICslaBuilder WithBlazorServerSupport(this ICslaBuilder builder)
    {
      builder.Services.AddTransient(typeof(ViewModel<>), typeof(ViewModel<>));
      builder.Services.AddSingleton<IAuthorizationPolicyProvider, CslaPermissionsPolicyProvider>();
      builder.Services.AddSingleton<IAuthorizationHandler, CslaPermissionsHandler>();
      return builder;
    }
```

Something in here must be the issue. 

The first line just sets up the `ViewModel` for DI, so that should be ok.

Then it sets up the Blazor authz objects. That must be the issue. I _suspect_ that perhaps there's no valid default user identity at this point (in other words, the default user might be `GenericPrincipal` or null, and it perhaps needs to be an unauthorized `ClaimsPrincipal`).

I'm not sure when I'll get time to do hands-on research. Can you try out my guess, @JacoJordaan, by setting `Csla.ApplicationContext.User` to a new `ClaimsPrincipal` right after the `UseCsla` call in your startup code?

**Reply by JacoJordaan** [reply]
@rockfordlhotka , looks like the issue is within `CslaPermissionsPolicyProvider`.

The fallback policy defaults to the default policy which requires an authenticated user.

I changed the following:
From: `public Task<AuthorizationPolicy> GetFallbackPolicyAsync() => FallbackPolicyProvider.GetDefaultPolicyAsync();`
To: `public Task<AuthorizationPolicy> GetFallbackPolicyAsync() => FallbackPolicyProvider.GetFallbackPolicyAsync();`

In `GetPolicyAsync` change the `return FallbackPolicyProvider.GetPolicyAsync(policyName);` to `return GetFallbackPolicyAsync();`
So, instead of
````
public Task<AuthorizationPolicy> GetPolicyAsync(string policyName)
    {
      if (CslaPolicy.TryGetPermissionRequirement(policyName, out CslaPermissionRequirement requirement))
      {
        var policy = new AuthorizationPolicyBuilder();
        policy.AddRequirements(requirement);
        return Task.FromResult(policy.Build());
      }
      else
      {
        return FallbackPolicyProvider.GetPolicyAsync(policyName);
      }
    }
````
To:
````
public Task<AuthorizationPolicy> GetPolicyAsync(string policyName)
        {
            if (CslaPolicy.TryGetPermissionRequirement(policyName, out CslaPermissionRequirement requirement))
            {
                var policy = new AuthorizationPolicyBuilder();
                policy.AddRequirements(requirement);
                return Task.FromResult(policy.Build());
            }
            else
            {
                return GetFallbackPolicyAsync();
            }
        }
````

A review of my logic is required, but it does solve the issue.

**Reply by JacoJordaan** [reply]
@rockfordlhotka, btw, I did try your suggestion by setting `Csla.ApplicationContext.User` to a new `ClaimsPrincipal` right after the `UseCsla` - no success. Only the change to `CslaPermissionsPolicyProvider` does the trick.

**Reply by rockfordlhotka** [reply]
Thank you for digging into the code, your solution looks reasonable to me.

I've logged a bug: https://github.com/MarimerLLC/csla/issues/1820

Do you want to submit a PR by any chance?

**Reply by JacoJordaan** [reply]
Thanks. I will do a PR.

**Comment by JacoJordaan** [comment]
The same issue when I run ProjectTracker with Blazor server.
You are able to display the Projects page if you are not logged in, but there are SignalR negotiation errors. If you have an interactive page then communication with the server will not work causing any events not updating the display.

![image](https://user-images.githubusercontent.com/43467554/94023179-326b7600-fdb6-11ea-8dd5-81ce66941a1d.png)
